### PR TITLE
Add documentation for Search Space

### PR DESCRIPTION
**Pull request by bouthilx**
Documents
- Priors
- Dimension types
- Special arguments (discrete, default_value, shape)
- Configuration (commandline, script's config file)
- Notes on transformations and conditional dimensions
- Give references to API

Link to compiled doc: https://orion.readthedocs.io/en/doc-search_space/user/searchspace.html

### CODE REVIEW

**Comment by bouthilx** [comment]
[Add documentation of config file](https://trello.com/c/8rI8Yoys/178-add-documentation-of-config-file)

**Comment by bouthilx** [comment]
[Add a `Search Space` page in the doc](https://trello.com/c/09v1aPuM/164-add-a-search-space-page-in-the-doc)

File: docs/src/user/searchspace.rst [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+The dimensions are casted to special types according to their prior. This is critical to
+either allow algorithms to leverage type information
+(ex: some algorithms works better on integers) or automatically transform trial types
+to make them compatible with specific algorithms
+(ex: some algorithms cannot work on categorical values).
+
+Real
+----
+
+All continous priors are automatically casted to :class:`Real`.
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: con·tin·u·ous

File: docs/src/user/searchspace.rst [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+Integer
+-------
+
+Discrete distributions of scipy are automatically casted to :class:`Integer`. All other
+distributions can be casted to :class:`Integer` by setting ``discrete=True``
+(ex: ``uniform(low, high, discrete=True)``).
+
+.. warning::
+
+   We recommend using continous priors with ``discrete=True``
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: continuous

File: docs/src/user/searchspace.rst [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+   as there is an issue with scipy discrete distribution because of incorrect interval. Issue
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: intervals*

File: docs/src/user/searchspace.rst [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+Special placeholder to represent a fidelity dimension. Only the prior
+:ref:`prior-fidelity` is casted to :class:`Fidelity`.
+
+Special arguments
+=================
+
+``discrete``
+------------
+
+ex: ``unifor(0, 10, discrete=True)``
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: uniform*

File: docs/src/user/searchspace.rst [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+Argument to cast a continous distribution into :ref:`integer-dim`. Defaults to ``False``.
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: continuous

File: docs/src/user/searchspace.rst [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+``default_value``
+-----------------
+
+ex: ``unifor(0, 10, default_value=5)``
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: Uniform*

File: docs/src/user/searchspace.rst [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+
+Dimensions can be set to a default value so that commandline call `insert` can support insertion
+without specifing this hyperparameter, assigning it the default value. This is also usefull in when
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: Specifying*, `useful when`*

File: docs/src/user/searchspace.rst [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+using the :ref:`EVC system`, so that experiments where an hyperparameter is deleted or added can
+adapt trials from other experiments by using the default value.
+
+``shape``
+---------
+
+ex: ``unifor(0, 10, shape=2)``
+
+Some hyper-parameters may have multiple dimensions. This can be set using ``shape``
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: Hyperparameters*

File: docs/src/user/searchspace.rst [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+Dimensions can be set to a default value so that commandline call `insert` can support insertion
+without specifing this hyperparameter, assigning it the default value. This is also usefull in when
+using the :ref:`EVC system`, so that experiments where an hyperparameter is deleted or added can
+adapt trials from other experiments by using the default value.
+
+``shape``
+---------
+
+ex: ``unifor(0, 10, shape=2)``
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: uniform*

File: docs/src/user/searchspace.rst [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+Configuration file
+-------------------
+
+You can use configuration files to define search space with placeholder
+``'orion~dist(*args, **kwargs)'`` in yaml and json files or
+``name~dist(*args, **kwargs)`` in any other text-based file.
+For now Oríon can only recognize the
+configuration file if it is passed with the argument ``--config`` to the user script. This should
+not be confused with the argument ``--config`` of ``orion hunt``, which is the configuration of
+Oríon. We are here referring the configuration of the user script, represented with
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: to the*

File: docs/src/user/searchspace.rst [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+
+    {
+      "lr": "orion~loguniform(1e-5, 1.0)"
+      "model": {
+        "activation": "orion~choices(['sigmoid', 'relu'])"
+        "hiddens": "orion~randint(100, 1000)"
+      }
+    }
+
+And here is an example with python! Note that for other files than for json and yaml, the
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: Note that for files other than

File: docs/src/user/searchspace.rst [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+- :class:`Real` are quantized to :class:`Integer`.
+- :class:`Categorical` are casted to :class:`Integer` (low=0, high=number of categories).
+  (probabilities are lost if defined)
+
+Conditional dependencies
+------------------------
+
+There is currently no support for conditional dependencies between dimensions.
+Conditional dependencies arises in situations where some hyperparameter defines which algorithm to
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: arise*

File: docs/src/user/searchspace.rst [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+use and each algorithm have its own
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: has its own

File: docs/src/user/searchspace.rst [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+set of different hyperparameter. We plan to support this in the future by replacing our current
```
Comment (DatCorno) [inline docs/src/user/searchspace.rst]: hyperparameters

**Pull request merged**